Modules/Microsoft.PowerShell.UnixCompleters/Microsoft.PowerShell.UnixCompleters/UtilCompleterInitializer.cs

```diff
@@ -0,0 +1,133 @@
+
+        private static Dictionary<string, ScriptBlock> s_nativeArgumentCompleterTable;
+
+        internal static IEnumerable<string> CompletedCommands { get; set; }
+
+        internal static Dictionary<string, ScriptBlock> NativeArgumentCompleterTable
+        {
+            get
+            {
+                if (s_nativeArgumentCompleterTable == null)
+                {
+                    object executionContext = s_executionContextProperty.GetValue(Runspace.DefaultRunspace);
+                    
+                    var completerTable = (Dictionary<string, ScriptBlock>)s_nativeArgumentCompletersProperty.GetValue(executionContext);
+
+                    if (completerTable == null)
+                    {
+                        completerTable = new Dictionary<string, ScriptBlock>(StringComparer.OrdinalIgnoreCase);
+                        s_nativeArgumentCompletersProperty.SetValue(executionContext, completerTable);
+                    }
+
+                    s_nativeArgumentCompleterTable = completerTable;
+                }
+
+                return s_nativeArgumentCompleterTable;
+            }
+        }
+
+        internal static IUnixUtilCompleter UnixUtilCompleter { get; set; }
+    }
+
+    public class UtilCompleterInitializer : IModuleAssemblyInitializer
+    {
+        private const string SHELL_PREFERENCE_VARNAME = "COMPLETION_SHELL_PREFERENCE";
+
+        public void OnImport()
+        {
+            string preferredCompletionShell = Environment.GetEnvironmentVariable(SHELL_PREFERENCE_VARNAME);
+
+            ShellType shellType = ShellType.None;
+            string shellExePath;
+            if ((string.IsNullOrEmpty(preferredCompletionShell) || !UnixHelpers.TryFindShell(preferredCompletionShell, out shellExePath, out shellType))
+                && !UnixHelpers.TryFindFallbackShell(out shellExePath, out shellType))
+            {
+                WriteError("Unable to find shell to provide unix utility completions");
+                return;
+            }
+
+            IUnixUtilCompleter utilCompleter;
+            switch (shellType)
+            {
+                case ShellType.Bash:
+                    utilCompleter = new BashUtilCompleter(shellExePath);
+                    break;
+
+                case ShellType.Zsh:
+                    utilCompleter = new ZshUtilCompleter(shellExePath);
+                    break;
+
+                default:
+                    WriteError("Unable to find shell to provide unix utility completions");
+                    return;
+            }
+
+            IEnumerable<string> utilsToComplete = utilCompleter.FindCompletableCommands();
+
+            CompleterGlobals.CompletedCommands = utilsToComplete;
+            CompleterGlobals.UnixUtilCompleter = utilCompleter;
+
+            RegisterCompletersForCommands(utilsToComplete);
```

Comment:
Will there be a problem if the module gets imported to two Runspaces? namely, `Register-ArgumentCompleter` would be called twice for each of the same native utility.

Reply:
Quite likely. I don't think it's threadsafe either (nor is the underlying mechanism for `Register-ArgumentCompleter`; it's not a `ConcurrentDictionary<,>`. But as a module intended solely for interactive scenarios, I'm thinking of multithreaded/multirunspace scenarios as out of scope for now.

Reply:
That's a fair point.